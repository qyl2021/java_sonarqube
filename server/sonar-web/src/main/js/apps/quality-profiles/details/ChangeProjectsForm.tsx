/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { find, without } from 'lodash';
import * as React from 'react';
import {
  associateProject,
  dissociateProject,
  getProfileProjects,
  ProfileProject
} from '../../../api/quality-profiles';
import Modal from '../../../components/controls/Modal';
import SelectList, { Filter } from '../../../components/SelectList/SelectList';
import { translate } from '../../../helpers/l10n';
import { Profile } from '../types';

interface Props {
  onClose: () => void;
  organization: string | null;
  profile: Profile;
}

export interface SearchParams {
  key: string;
  organization: string | null;
  page: number;
  pageSize: number;
  query?: string;
  selected: string;
}

interface State {
  lastSearchParams: SearchParams;
  listHasBeenTouched: boolean;
  projects: ProfileProject[];
  projectsTotalCount?: number;
  selectedProjects: string[];
}

const PAGE_SIZE = 100;

export default class ChangeProjectsForm extends React.PureComponent<Props, State> {
  mounted = false;

  constructor(props: Props) {
    super(props);

    this.state = {
      lastSearchParams: {
        key: props.profile.key,
        organization: props.organization,
        page: 1,
        pageSize: PAGE_SIZE,
        query: '',
        selected: Filter.Selected
      },
      listHasBeenTouched: false,
      projects: [],
      selectedProjects: []
    };
  }

  componentDidMount() {
    this.mounted = true;
    this.fetchProjects(this.state.lastSearchParams);
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  fetchProjects = (searchParams: SearchParams, more?: boolean) =>
    getProfileProjects({
      ...searchParams,
      p: searchParams.page,
      ps: searchParams.pageSize,
      q: searchParams.query !== '' ? searchParams.query : undefined
    }).then(data => {
      if (this.mounted) {
        this.setState(prevState => {
          const projects = more ? [...prevState.projects, ...data.results] : data.results;
          const newSeletedProjects = data.results
            .filter(project => project.selected)
            .map(project => project.key);
          const selectedProjects = more
            ? [...prevState.selectedProjects, ...newSeletedProjects]
            : newSeletedProjects;

          return {
            lastSearchParams: searchParams,
            listHasBeenTouched: false,
            projects,
            projectsTotalCount: data.paging.total,
            selectedProjects
          };
        });
      }
    });

  handleLoadMore = () =>
    this.fetchProjects(
      {
        ...this.state.lastSearchParams,
        page: this.state.lastSearchParams.page + 1
      },
      true
    );

  handleReload = () =>
    this.fetchProjects({
      ...this.state.lastSearchParams,
      page: 1
    });

  handleSearch = (query: string, selected: Filter) =>
    this.fetchProjects({
      ...this.state.lastSearchParams,
      page: 1,
      query,
      selected
    });

  handleSelect = (key: string) =>
    associateProject(this.props.profile.key, key).then(() => {
      if (this.mounted) {
        this.setState((state: State) => ({
          listHasBeenTouched: true,
          selectedProjects: [...state.selectedProjects, key]
        }));
      }
    });

  handleUnselect = (key: string) =>
    dissociateProject(this.props.profile.key, key).then(() => {
      if (this.mounted) {
        this.setState((state: State) => ({
          listHasBeenTouched: true,
          selectedProjects: without(state.selectedProjects, key)
        }));
      }
    });

  handleCloseClick = (event: React.SyntheticEvent<HTMLElement>) => {
    event.preventDefault();
    this.props.onClose();
  };

  renderElement = (key: string): React.ReactNode => {
    const project = find(this.state.projects, { key });
    return project === undefined ? key : project.name;
  };

  render() {
    const header = translate('projects');

    return (
      <Modal contentLabel={header} onRequestClose={this.props.onClose}>
        <div className="modal-head">
          <h2>{header}</h2>
        </div>

        <div className="modal-body modal-container" id="profile-projects">
          <SelectList
            allowBulkSelection={true}
            elements={this.state.projects.map(project => project.key)}
            elementsTotalCount={this.state.projectsTotalCount}
            labelAll={translate('quality_gates.projects.all')}
            labelSelected={translate('quality_gates.projects.with')}
            labelUnselected={translate('quality_gates.projects.without')}
            needReload={
              this.state.listHasBeenTouched && this.state.lastSearchParams.selected !== Filter.All
            }
            onLoadMore={this.handleLoadMore}
            onReload={this.handleReload}
            onSearch={this.handleSearch}
            onSelect={this.handleSelect}
            onUnselect={this.handleUnselect}
            renderElement={this.renderElement}
            selectedElements={this.state.selectedProjects}
          />
        </div>

        <div className="modal-foot">
          <a href="#" onClick={this.handleCloseClick}>
            {translate('close')}
          </a>
        </div>
      </Modal>
    );
  }
}
