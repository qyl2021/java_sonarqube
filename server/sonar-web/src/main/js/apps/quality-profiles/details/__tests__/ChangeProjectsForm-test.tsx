/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import * as React from 'react';
import ChangeProjectsForm, { SearchParams } from '../ChangeProjectsForm';
import SelectList, { Filter } from '../../../../components/SelectList/SelectList';
import { waitAndUpdate } from '../../../../helpers/testUtils';
import {
  getProfileProjects,
  associateProject,
  dissociateProject
} from '../../../../api/quality-profiles';

jest.mock('../../../../api/quality-profiles', () => ({
  getProfileProjects: jest.fn().mockResolvedValue({
    paging: { pageIndex: 1, pageSize: 3, total: 55 },
    results: [
      { id: 'test1', key: 'test1', name: 'test1', selected: false },
      { id: 'test2', key: 'test2', name: 'test2', selected: false },
      { id: 'test3', key: 'test3', name: 'test3', selected: true }
    ]
  }),
  associateProject: jest.fn().mockResolvedValue({}),
  dissociateProject: jest.fn().mockResolvedValue({})
}));

const profile: any = { key: 'profFile_key' };

beforeEach(() => {
  jest.clearAllMocks();
});

it('should render correctly', async () => {
  const wrapper = shallowRender();
  await waitAndUpdate(wrapper);

  expect(wrapper).toMatchSnapshot();
  expect(getProfileProjects).toHaveBeenCalled();

  wrapper.setState({ listHasBeenTouched: true });
  expect(wrapper.find(SelectList).props().needReload).toBe(true);

  wrapper.setState({ lastSearchParams: { selected: Filter.All } as SearchParams });
  expect(wrapper.find(SelectList).props().needReload).toBe(false);
});

it('should handle reload properly', async () => {
  const wrapper = shallowRender();
  await waitAndUpdate(wrapper);

  wrapper.instance().handleReload();
  expect(getProfileProjects).toHaveBeenCalledWith(
    expect.objectContaining({
      p: 1
    })
  );
  expect(wrapper.state().listHasBeenTouched).toBe(false);
});

it('should handle search reload properly', async () => {
  const wrapper = shallowRender();
  await waitAndUpdate(wrapper);

  wrapper.instance().handleSearch('foo', Filter.Selected);
  expect(getProfileProjects).toHaveBeenCalledWith(
    expect.objectContaining({
      p: 1,
      q: 'foo',
      selected: Filter.Selected
    })
  );
  expect(wrapper.state().listHasBeenTouched).toBe(false);
});

it('should handle load more properly', async () => {
  const wrapper = shallowRender();
  await waitAndUpdate(wrapper);

  wrapper.instance().handleLoadMore();
  expect(getProfileProjects).toHaveBeenCalledWith(
    expect.objectContaining({
      p: 2
    })
  );
  expect(wrapper.state().listHasBeenTouched).toBe(false);
});

it('should handle selection properly', async () => {
  const wrapper = shallowRender();
  await waitAndUpdate(wrapper);

  wrapper.instance().handleSelect('toto');
  await waitAndUpdate(wrapper);
  expect(associateProject).toHaveBeenCalledWith(profile.key, 'toto');
  expect(wrapper.state().listHasBeenTouched).toBe(true);
});

it('should handle deselection properly', async () => {
  const wrapper = shallowRender();
  await waitAndUpdate(wrapper);

  wrapper.instance().handleUnselect('tata');
  await waitAndUpdate(wrapper);
  expect(dissociateProject).toHaveBeenCalledWith(profile.key, 'tata');
  expect(wrapper.state().listHasBeenTouched).toBe(true);
});

function shallowRender(props: Partial<ChangeProjectsForm['props']> = {}) {
  return shallow<ChangeProjectsForm>(
    <ChangeProjectsForm onClose={jest.fn()} organization={'TEST'} profile={profile} {...props} />
  );
}
