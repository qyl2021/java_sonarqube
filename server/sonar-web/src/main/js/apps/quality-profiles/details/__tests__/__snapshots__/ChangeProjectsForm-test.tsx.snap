// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<Modal
  contentLabel="projects"
  onRequestClose={[MockFunction]}
>
  <div
    className="modal-head"
  >
    <h2>
      projects
    </h2>
  </div>
  <div
    className="modal-body modal-container"
    id="profile-projects"
  >
    <SelectList
      allowBulkSelection={true}
      elements={
        Array [
          "test1",
          "test2",
          "test3",
        ]
      }
      elementsTotalCount={55}
      labelAll="quality_gates.projects.all"
      labelSelected="quality_gates.projects.with"
      labelUnselected="quality_gates.projects.without"
      needReload={false}
      onLoadMore={[Function]}
      onReload={[Function]}
      onSearch={[Function]}
      onSelect={[Function]}
      onUnselect={[Function]}
      renderElement={[Function]}
      selectedElements={
        Array [
          "test3",
        ]
      }
    />
  </div>
  <div
    className="modal-foot"
  >
    <a
      href="#"
      onClick={[Function]}
    >
      close
    </a>
  </div>
</Modal>
`;
